### ISSUE TITLE
2 test failed in x86_checked_windows_nt_gcstress0x3

### ISSUE THREAD

**Issue by AriNuer** [dotnet/coreclr#20689]
**job**
https://ci.dot.net/job/dotnet_coreclr/job/master/view/x86/job/jitstress/job/x86_checked_windows_nt_gcstress0x3/101/#showFailuresLink

**failed tests**
Interop_PInvoke._Delegate_MarshalDelegateAsField_AsFunctionPtr_AsFunctionPtrTest_AsFunctionPtrTest_._Delegate_MarshalDelegateAsField_AsFunctionPtr_AsFunctionPtrTest_AsFunctionPtrTest_cmd
Interop_PInvoke._Delegate_MarshalDelegateAsField_AsDefault_AsDefaultTest_AsDefaultTest_._Delegate_MarshalDelegateAsField_AsDefault_AsDefaultTest_AsDefaultTest_cmd
**log:**
```
Interop_PInvoke._Delegate_MarshalDelegateAsField_AsFunctionPtr_AsFunctionPtrTest_AsFunctionPtrTest_._Delegate_MarshalDelegateAsField_AsFunctionPtr_AsFunctionPtrTest_AsFunctionPtrTest_cmd [FAIL]
        
  
  Return code:      1
  Raw output file:      D:\j\workspace\x86_checked_w---71335245\bin\tests\Windows_NT.x86.Checked\Reports\Interop.PInvoke\Delegate\MarshalDelegateAsField\AsFunctionPtr\AsFunctionPtrTest\AsFunctionPtrTest.output.txt
  Raw output:
  BEGIN EXECUTION
         "D:\j\workspace\x86_checked_w---71335245\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root\corerun.exe" AsFunctionPtrTest.exe 
        Scenario 1 : Delegate marshaled as field in struct with Sequential.
        
  
  Scenario 2 : Delegate marshaled as field in struct with Explicit.
        
  
  Scenario 3 : Delegate marshaled as field in class with Sequential.
        Expected: 100
        Actual: -1073741819
        END EXECUTION - FAILED
        FAILED
        Test Harness Exitcode is : 1
        
  To run the test:
  > set CORE_ROOT=D:\j\workspace\x86_checked_w---71335245\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root
  > D:\j\workspace\x86_checked_w---71335245\bin\tests\Windows_NT.x86.Checked\Interop\PInvoke\Delegate\MarshalDelegateAsField\AsFunctionPtr\AsFunctionPtrTest\AsFunctionPtrTest.cmd
  
        Expected: True
        Actual:   False
        Stack Trace:
          D:\j\workspace\x86_checked_w---71335245\bin\tests\Windows_NT.x86.Checked\TestWrappers\Interop.PInvoke\Interop.PInvoke.XUnitWrapper.cs(249,0): at Interop_PInvoke._Delegate_MarshalDelegateAsField_AsFunctionPtr_AsFunctionPtrTest_AsFunctionPtrTest_._Delegate_MarshalDelegateAsField_AsFunctionPtr_AsFunctionPtrTest_AsFunctionPtrTest_cmd()
```

```
Interop_PInvoke._Delegate_MarshalDelegateAsField_AsDefault_AsDefaultTest_AsDefaultTest_._Delegate_MarshalDelegateAsField_AsDefault_AsDefaultTest_AsDefaultTest_cmd [FAIL]
        
  
  Return code:      1
  Raw output file:      D:\j\workspace\x86_checked_w---71335245\bin\tests\Windows_NT.x86.Checked\Reports\Interop.PInvoke\Delegate\MarshalDelegateAsField\AsDefault\AsDefaultTest\AsDefaultTest.output.txt
  Raw output:
  BEGIN EXECUTION
         "D:\j\workspace\x86_checked_w---71335245\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root\corerun.exe" AsDefaultTest.exe 
        Scenario 1 : Delegate marshaled as field in struct with Sequential.
        
  
  Scenario 2 : Delegate marshaled as field in struct with Explicit.
        
  
  Scenario 3 : Delegate marshaled as field in class with Sequential.
        Expected: 100
        Actual: -1073741819
        END EXECUTION - FAILED
        FAILED
        Test Harness Exitcode is : 1
        
  To run the test:
  > set CORE_ROOT=D:\j\workspace\x86_checked_w---71335245\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root
  > D:\j\workspace\x86_checked_w---71335245\bin\tests\Windows_NT.x86.Checked\Interop\PInvoke\Delegate\MarshalDelegateAsField\AsDefault\AsDefaultTest\AsDefaultTest.cmd
  
        Expected: True
        Actual:   False
        Stack Trace:
          D:\j\workspace\x86_checked_w---71335245\bin\tests\Windows_NT.x86.Checked\TestWrappers\Interop.PInvoke\Interop.PInvoke.XUnitWrapper.cs(179,0): at Interop_PInvoke._Delegate_MarshalDelegateAsField_AsDefault_AsDefaultTest_AsDefaultTest_._Delegate_MarshalDelegateAsField_AsDefault_AsDefaultTest_AsDefaultTest_cmd()
```

**Comment by sandreenko** [comment]
Tests were added in #19323

**Comment by sandreenko** [comment]
The repro needs checked build and gcstress=0x2.

**Comment by sandreenko** [comment]
Fails with:
```
Assert failure(PID 22016 [0x00005600], Thread: 34132 [0x8554]): CONTRACT VIOLATION by UMEntryThunk::ReportViolation at "f:\git\coreclr\src\vm\dllimportcallback.cpp" @ 1021

MODE_COOPERATIVE encountered while thread is in preemptive state.

                        CONTRACT in UMEntryThunk::ReportViolation at "f:\git\coreclr\src\vm\dllimportcallback.cpp" @ 1021
                        CONTRACT in MethodDescCallSite::CallTargetWorker at "f:\git\coreclr\src\vm\callhelpers.cpp" @ 367
                        GCX_COOP in Assembly::ExecuteMainMethod at "f:\git\coreclr\src\vm\assembly.cpp" @ 1760
                        CONTRACT in Assembly::ExecuteMainMethod at "f:\git\coreclr\src\vm\assembly.cpp" @ 1744
                        GCX_COOP in CorHost2::ExecuteAssembly at "f:\git\coreclr\src\vm\corhost.cpp" @ 469
                        CONTRACT in CorHost2::ExecuteAssembly at "f:\git\coreclr\src\vm\corhost.cpp" @ 413

We can't find the violated contract. Look for an old-style non-holder-based contract.


CORECLR! CONTRACT_ASSERT + 0x37B (0x0fd9dd4b)
CORECLR! EEContract::DoChecks + 0x1AA (0x0feb020a)
CORECLR! UMEntryThunk::ReportViolation + 0x195 (0x10297325)
PINVOKE_DELEGATE_ASFIELD! TakeDelegateAsFieldInClass_Seq + 0x1D (0x587410fd)
<no module>! <no symbol> + 0x0 (0x00464965)
<no module>! <no symbol> + 0x0 (0x00462af2)
CORECLR! CallDescrWorkerInternal + 0x34 (0x0feb0831)
CORECLR! CallDescrWorker + 0x124 (0x1028eb64)
CORECLR! CallDescrWorkerWithHandler + 0x205 (0x1028edf5)
CORECLR! MethodDescCallSite::CallTargetWorker + 0x974 (0x1028f7b4)
    File: f:\git\coreclr\src\vm\dllimportcallback.cpp Line: 1021
    Image: F:\git\coreclr\bin\tests\Windows_NT.x86.Debug\Tests\Core_Root\CoreRun.exe
```

**Comment by jkoritzinsky** [comment]
Fixed by #20896.
